src/rest_vol.c

```diff
@@ -3794,11 +3796,60 @@ RV_curl_multi_perform(CURL *curl_multi_handle, dataset_transfer_info *transfer_i
                     transfer_info[handle_index].curl_easy_handle = NULL;
 
                     if (transfer_info[handle_index].transfer_type == WRITE) {
-                        RV_free(transfer_info[handle_index].u.write_info.write_body);
-                        transfer_info[handle_index].u.write_info.write_body = NULL;
+                        if (transfer_info[handle_index].tconv_buf) {
+                            if ((dtype_class = H5Tget_class(transfer_info[handle_index].mem_type_id)) ==
+                                H5T_NO_CLASS)
+                                FUNC_DONE_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't get mem dtype class");
+
+                            /* Clean up memory allocated by type conversion of vlen types */
+                            if (dtype_class == H5T_VLEN) {
```

Comment:
This may become complicated for things like compound datatypes with variable-length fields inside. You probably want to use `H5Tdetect_class(..., H5T_VLEN)` for this.